pkg/networkservice/common/onidle/server.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) 2021 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package onidle provides server chain element that executes a callback when there were no connections for specified time
+package onidle
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/pkg/errors"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/clock"
+)
+
+type onIdleServer struct {
+	ctx         context.Context
+	timeout     time.Duration
+	notify      func()
+	activeConns sync.Map
+	timer       clock.Timer
+	// Timers don't support concurrency natively.
+	// If we stop it, they tell us if the timer was running before our call.
+	// But if timer was not running, there's no way to distinguish
+	// if it was stopped earlier (e.g. concurrently by another thread) or if it has already fired.
+	// Therefore, we should implement some manual sync for this.
+	timerMut     sync.Mutex
+	timerFired   bool
+	timerCounter int
+}
+
+// NewServer - returns a new server chain element that notifies about long time periods without active requests
+func NewServer(ctx context.Context, notify func(), options ...Option) networkservice.NetworkServiceServer {
+	clockTime := clock.FromContext(ctx)
+
+	t := &onIdleServer{
+		ctx:     ctx,
+		timeout: time.Minute * 10,
+		notify:  notify,
+	}
+
+	for _, opt := range options {
+		opt(t)
+	}
+
+	t.timer = clockTime.Timer(t.timeout)
```

Comment:
Probably `clockTime.AfterFunc()` would be a better solution?
```go
timer := clockTime.Timer(timeout)
go func() {
    for {
        select {
        case <-ctx.Done():
            t.timer.Close()
            return
        case <-timer.C()
            doTimerJob()
    }
}
``` 
Can be replaced with:
```go
timer := clockTime.Timer(timeout, doTimerJob)
go func() {
    <-ctx.Done()
    t.timer.Close()
}
```